js.go

```diff
@@ -0,0 +1,1246 @@
+	// JSApiConsumerCreateT is used to create consumers.
+	JSApiConsumerCreateT = "CONSUMER.CREATE.%s"
+	// JSApiDurableCreateT is used to create durable consumers.
+	JSApiDurableCreateT = "CONSUMER.DURABLE.CREATE.%s.%s"
+	// JSApiConsumerInfoT is used to create consumers.
+	JSApiConsumerInfoT = "CONSUMER.INFO.%s.%s"
+	// JSApiRequestNextT is the prefix for the request next message(s) for a consumer in worker/pull mode.
+	JSApiRequestNextT = "CONSUMER.MSG.NEXT.%s.%s"
+	// JSApiStreamCreate is the endpoint to create new streams.
+	JSApiStreamCreateT = "STREAM.CREATE.%s"
+)
+
+// JetStream returns a JetStream context for pub/sub interactions.
+func (nc *Conn) JetStream(opts ...JSOpt) (JetStream, error) {
+	const defaultRequestWait = 5 * time.Second
+
+	js := &js{nc: nc, pre: JSDefaultApiPrefix, wait: defaultRequestWait}
+
+	for _, f := range opts {
+		if err := f(js); err != nil {
+			return nil, err
+		}
+	}
+
+	if js.direct {
+		return js, nil
+	}
+
+	resp, err := nc.Request(js.apiSubj(JSApiAccountInfo), nil, js.wait)
+	if err != nil {
+		if err == ErrNoResponders {
+			err = ErrJetStreamNotEnabled
+		}
+		return nil, err
+	}
+	var info AccountInfoResponse
+	if err := json.Unmarshal(resp.Data, &info); err != nil {
+		return nil, err
+	}
+	if info.Error != nil && info.Error.Code == 503 {
+		return nil, ErrJetStreamNotEnabled
+	}
+	return js, nil
+}
+
+// JSOpt configures options for the jetstream context.
+type JSOpt func(opts *js) error
+
+func ApiPrefix(pre string) JSOpt {
+	return func(js *js) error {
+		js.pre = pre
+		if !strings.HasSuffix(js.pre, ".") {
+			js.pre = js.pre + "."
+		}
+		return nil
+	}
+}
+
+func ApiRequestWait(wait time.Duration) JSOpt {
+	return func(js *js) error {
+		js.wait = wait
+		return nil
+	}
+}
+
+func DirectOnly() JSOpt {
+	return func(js *js) error {
+		js.direct = true
+		return nil
+	}
+}
+
+func (js *js) apiSubj(subj string) string {
+	if js.pre == _EMPTY_ {
+		return subj
+	}
+	var b strings.Builder
+	b.WriteString(js.pre)
+	b.WriteString(subj)
+	return b.String()
+}
+
+// PubOpt configures options for publishing jetstream messages.
+type PubOpt func(opts *pubOpts) error
+
+type pubOpts struct {
+	ctx context.Context
+	ttl time.Duration
+	id  string
+	lid string // Expected last msgId
+	str string // Expected stream name
+	seq uint64 // Expected last sequence
+}
+
+type PubAckResponse struct {
+	ApiResponse
+	*PubAck
+}
+
+type PubAck struct {
+	Stream    string `json:"stream"`
+	Sequence  uint64 `json:"seq"`
+	Duplicate bool   `json:"duplicate,omitempty"`
+}
+
+// Headers for published messages.
+const (
+	MsgIdHdr             = "Nats-Msg-Id"
+	ExpectedStreamHdr    = "Nats-Expected-Stream"
+	ExpectedLastSeqHdr   = "Nats-Expected-Last-Sequence"
+	ExpectedLastMsgIdHdr = "Nats-Expected-Last-Msg-Id"
+)
+
+func (js *js) PublishMsg(m *Msg, opts ...PubOpt) (*PubAck, error) {
+	var o pubOpts
+	if len(opts) > 0 {
+		if m.Header == nil {
+			m.Header = http.Header{}
+		}
+		for _, f := range opts {
```

Comment:
Have we evaluated the performance effect of variadic publish options as opposed to passing a pointer to a structure? It may be fine for a connect, or subscribe, but I wonder if this is a good approach for publish due to the fact that it can be called many times a second ;-).

Reply:
Publish is an RPC, none of that matters TBH. I care about costs of processing in the server though for sure.